### PR TITLE
Fix S3 FileReader when using repeated samples (pad_last_batch=True)

### PR DESCRIPTION
**Pull request by jantonguirao**
## Category:

**Bug fix** 

## Description:

- Fixes an issue with the FileReader when using S3 URIs in combination with pad_last_batch=True.
- The bug comes when the file reading happens in parallel, and two threads try to open/read/close the same file stream at the same time (as it happens when we pad the last batch with repeated samples)
- Fixes https://github.com/NVIDIA/DALI/issues/1281#issuecomment-2142425540

### Affected modules and functionalities:

FileReader

### Tests:

- [ ] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [x] N/A
Tested manually 

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-3975

### CODE REVIEW

File: dali/operators/reader/file_reader_op.h [hunk 1]
```diff
@@ -61,31 +63,52 @@ class FileReader : public DataReader<CPUBackend, ImageLabelWrapper, ImageLabelWr
...
 
   void RunImpl(Workspace &ws) override {
-    auto &image_output = ws.Output<CPUBackend>(0);
+    auto &file_output = ws.Output<CPUBackend>(0);
     auto &label_output = ws.Output<CPUBackend>(1);
     auto samples = GetCurrBatch();
     int batch_size = samples.size();
 
     auto &thread_pool = ws.GetThreadPool();
+    std::unordered_map<void *, SmallVector<int, 4>> repeated_samples;
```
Comment (JanuszL) [inline dali/operators/reader/file_reader_op.h:72]: How about 
```suggestion
    std::unordered_map<void *, SmallVector<int, 4>> unique_samples;
```

**Review by JanuszL — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

**Comment by szkarpinski** [comment]
Would it be possible to add the repro as a test to confirm that the issue is gone?

**Comment by jantonguirao** [comment]
> Would it be possible to add the repro as a test to confirm that the issue is gone?

That would require setting some S3 hosting infrastructure for it. Out of scope for this PR, but I think we should invest in it at some point

File: dali/operators/reader/file_reader_op.h [hunk 2]
```diff
@@ -61,31 +63,52 @@ class FileReader : public DataReader<CPUBackend, ImageLabelWrapper, ImageLabelWr
...
+    auto &file_output = ws.Output<CPUBackend>(0);
     auto &label_output = ws.Output<CPUBackend>(1);
     auto samples = GetCurrBatch();
     int batch_size = samples.size();
 
     auto &thread_pool = ws.GetThreadPool();
+    std::unordered_map<void *, SmallVector<int, 4>> unique_samples;
     for (int sample_idx = 0; sample_idx < batch_size; ++sample_idx) {
-      thread_pool.AddWork([&, sample_idx](int tid) {
+      auto sample_ptr = samples[sample_idx].get();
```
Comment (klecki) [inline dali/operators/reader/file_reader_op.h:74]: Nitpick:
```suggestion
      auto *sample_ptr = samples[sample_idx].get();
```

**Review by klecki — APPROVED** [review]

**Review by szkarpinski — APPROVED** [review]

**Pull request merged**